platform-sdk/swirlds-filesystem-manager/src/main/java/com/swirlds/filesystem/manager/internal/FileSystemManagerImpl.java

```diff
@@ -0,0 +1,96 @@
+ */
+
+package com.swirlds.filesystem.manager.internal;
+
+import com.swirlds.common.io.utility.RecycleBin;
+import com.swirlds.filesystem.manager.FileSystemManager;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class FileSystemManagerImpl implements FileSystemManager {
+
+    private final Path rootPath;
+    private final RecycleBin bin;
+    private static final AtomicLong FILE_ID = new AtomicLong(0);
+
+    FileSystemManagerImpl(@NonNull final Path rootPath, @NonNull final RecycleBin bin) {
+        this.rootPath = rootPath;
+        this.bin = bin;
+    }
+
+    /**
+     * {@inheritDoc}
+     * This implementation is nominal.
+     */
+    @NonNull
+    @Override
+    public Path resolve(@NonNull final Path relativePath) {
+        return subPathFromTempLocation(relativePath);
+    }
+
+    /**
+     * Creates a temporary directory relative to the root directory of the file system manager.
+     *
+     * A class level index is used in order to assure uniqueness of the files in the directory.
+     *
+     * @param name the name to use as suffix for the directory
+     * @return the resolved path
+     * @throws IllegalArgumentException if the path cannot be relative to or scape the root directory
+     */
+    @NonNull
+    @Override
+    public Path createTemporaryPath(@NonNull final String name) {
+        final String fileName = FILE_ID.getAndIncrement() + "-" + name;
+        return subPathFromTempLocation(Paths.get(fileName));
+    }
+
+    /**
+     * {@inheritDoc}
+     * This implementation is nominal.
+     */
+    @Override
+    public void recycle(@NonNull final Path relativePath) throws IOException {
+        bin.recycle(subPathFromTempLocation(relativePath));
+    }
+
+    private Path subPathFromTempLocation(@NonNull final Path subpath) {
+        final Path normalizedSubpath = // Remove redundant name parts
+                requireNotSameAsTempLocation(rootPath.resolve(subpath).normalize());
+        final Path relativePath = rootPath.relativize(normalizedSubpath);
+        // Check if relativization is within temporaryFileLocation
+        if (relativePath.startsWith("..")) {
+            throw new IllegalArgumentException("Requested path scape from temporary path '" + subpath + ".");
+        }
+        return normalizedSubpath;
+    }
+
+    private Path requireNotSameAsTempLocation(final @NonNull Path path) {
+        try {
+            if (Files.isSameFile(path, rootPath)) {
+                throw new IllegalArgumentException(
+                        "Requested subPath resolves to the root temporary path '" + path + ".");
+            }
+            return path;
+        } catch (IOException e) {
+            throw new IllegalArgumentException("Cannot assert if provided path against temporaryFileLocation", e);
+        }
+    }
```

Comment:
I suggest we do this slightly differently.

- root
    - temporary location
    - recycle bin location
    - data 

We should put all directories managed internally by the file system manager underneath the root. There should be a `data` directory (we can workshop the name for this if you don't like it), and all paths created by `resolve(path)` should be underneath the `data` directory. That way, there is never any possibility of collision between directories managed internally by the manager and directories managed by the end user of this utility.

Reply:
instead of data used usr, let me know what you thing about that

Reply:
As we spoke in slack, the data folder became the root of the fs manager, and saved became the user folder